redwood-lazylayout-compose/src/commonMain/kotlin/app/cash/redwood/lazylayout/compose/LazyListState.kt

```diff
@@ -18,32 +18,71 @@ package app.cash.redwood.lazylayout.compose
 import androidx.compose.runtime.Composable
 import androidx.compose.runtime.getValue
 import androidx.compose.runtime.mutableStateOf
-import androidx.compose.runtime.remember
+import androidx.compose.runtime.saveable.Saver
+import androidx.compose.runtime.saveable.rememberSaveable
 import androidx.compose.runtime.setValue
 
 @Composable
-public fun rememberLazyListState(
-  initialFirstVisibleItemIndex: Int = 0,
-): LazyListState {
-  return remember {
-    LazyListState(
-      initialFirstVisibleItemIndex,
-    )
+public fun rememberLazyListState(): LazyListState {
+  return rememberSaveable(saver = LazyListState.Saver) {
+    LazyListState()
   }
 }
 
-public class LazyListState(
-  firstVisibleItemIndex: Int = 0,
-) {
-  public var firstVisibleItemIndex: Int by mutableStateOf(firstVisibleItemIndex)
+public class LazyListState {
+  /** We only restore the scroll position once. */
+  private var hasRestoredScrollPosition = false
+
+  /** The scroll position to restore. */
+  private var restoredIndex: Int = -1
+
+  /**
+   * The value published to the host platform. This starts as 0 and changes exactly once to
+   * trigger exactly one scroll.
+   */
+  public var scrollItemIndex: Int by mutableStateOf(0)
     internal set
 
-  internal var scrollToItemTriggeredId by mutableStateOf(0)
+  public var firstVisibleItemIndex: Int = 0
+    private set
+
+  public fun restoreIndex(index: Int) {
```

Comment:
Could we change this back to `scrollToItem`? https://developer.android.com/reference/kotlin/androidx/compose/foundation/lazy/LazyListState#scrollToItem(kotlin.Int,kotlin.Int)

Reply:
this function is not really a scroll to item anymore, it has it's own logic, and what it does is "restore when there's criteria are met"